### PR TITLE
Requestify Witness Resolution

### PR DESCRIPTION
**Pull request by CodaFi**
Witness matching is a source of a lot of ad-hoc cycles, and mixes the
logic that performs resolution, caching, validation, and cycle breaking into one
place.  To make matters worse, some checkers kick off other checks in
order to cache work for further declarations, and access an internal
cache on their subject conformance for many requirements at once, or
sometimes just one requirement.

None of this fits into the request evaluator's central view of
caching.  This is further evidenced by the fact that if you attempt to
move the caching step into the evaluator, it overcaches the same
witness and trips asserts.

As a start, define requests for the resolution steps, and flush some
hacks around forcing witness resolution. The caching logic is mostly
untouched (the requests don't actually cache anything), but some cycle
breaking is now handled in the evaluator itself.  Once witness matching
has been refactored to cache with the evaluator, all of these hacks can
go away.

My urge to destroy the LazyResolver outweighs the compromises here.

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci please test

**Comment by CodaFi** [comment]
@swift-ci please test source compatibility

**Comment by CodaFi** [comment]
@DougGregor I *think* the right way to do this is to make a checker that acts like a pure function from a top-level target requirement and a (partial) `NormalConformance` to a `[Type]Witness`.  If further requirements need to get validated, then we can spawn sub-checkers with those requirements as targets.  Each checker would be responsible for caching *only* its target requirement on the conformance.

File: include/swift/AST/TypeCheckerTypeIDZone.def [hunk 1]
```diff
@@ -180,5 +182,10 @@ SWIFT_REQUEST(TypeChecker, HasDefaultInitRequest,
               bool(NominalTypeDecl *), Cached, NoLocationInfo)
 SWIFT_REQUEST(TypeChecker, SynthesizeDefaultInitRequest,
               ConstructorDecl *(NominalTypeDecl *), Cached, NoLocationInfo)
-SWIFT_REQUEST(TypeChecker, InheritsSuperclassInitializersRequest,
-              bool(ClassDecl *), SeparatelyCached, NoLocationInfo)
+SWIFT_REQUEST(TypeChecker, TypeWitnessRequest,
+              TypeWitnessAndDecl(NormalProtocolConformance *,
+                                 AssociatedTypeDecl *),
+              Cached, NoLocationInfo)
```
Comment (DougGregor) [inline include/swift/AST/TypeCheckerTypeIDZone.def]: You mean `SeparatelyCached` here?

File: include/swift/AST/TypeCheckerTypeIDZone.def [hunk 2]
```diff
@@ -180,5 +182,10 @@ SWIFT_REQUEST(TypeChecker, HasDefaultInitRequest,
...
+SWIFT_REQUEST(TypeChecker, ValueWitnessRequest,
+              Witness(NormalProtocolConformance *, ValueDecl *), Cached,
```
Comment (DougGregor) [inline include/swift/AST/TypeCheckerTypeIDZone.def]: ... and here?

File: lib/AST/ProtocolConformance.cpp [hunk 3]
```diff
@@ -1516,3 +1518,8 @@ const UnifiedStatsReporter::TraceFormatter*
 FrontendStatsTracer::getTraceFormatter<const ProtocolConformance *>() {
   return &TF;
 }
+
+void swift::simple_display(llvm::raw_ostream &out,
+                           const ProtocolConformance *conf) {
+  conf->dump(out);
```
Comment (DougGregor) [inline lib/AST/ProtocolConformance.cpp]: Should I worry about this causing other requests to fire?

File: lib/AST/TypeCheckRequests.cpp [hunk 4]
```diff
@@ -1071,3 +1072,37 @@ void InheritsSuperclassInitializersRequest::cacheResult(bool value) const {
...
+  auto *conformance = std::get<0>(getStorage());
+  auto *requirement = std::get<1>(getStorage());
+  if (conformance->TypeWitnesses.count(requirement) == 0) {
+    return None;
+  }
+  return conformance->TypeWitnesses[requirement];
+}
+
+void TypeWitnessRequest::cacheResult(TypeWitnessAndDecl typeWitAndDecl) const {
+  // FIXME: Refactor this to be the thing that warms the cache.
```
Comment (DougGregor) [inline lib/AST/TypeCheckRequests.cpp]: Yikes, but okay; I see your goal here.

**Review by DougGregor — APPROVED** [review]

File: include/swift/AST/TypeCheckerTypeIDZone.def (hunk 1 above)
Comment (CodaFi) [reply include/swift/AST/TypeCheckerTypeIDZone.def]: Whoops.

File: lib/AST/ProtocolConformance.cpp (hunk 3 above)
Comment (CodaFi) [reply lib/AST/ProtocolConformance.cpp]: This is a placeholder... I'll fix it up.

Comment (CodaFi) [reply lib/AST/ProtocolConformance.cpp]: Oh wow, the printer can kick off interface type requests for var decls will can perform arbitrarily large amounts of type checking.  Lovely.

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Review by slavapestov — APPROVED** [review]

**Review by DougGregor — APPROVED** [review]

**Pull request closed without merging**